Pass IsInteractive context to assembly reader shim

auduchinok opened the pull request:
When reading assemblies via AssemblyReaderShim there's currently no way to distinguish between using these assemblies from F# projects and from scripting/interactive environment. This knowledge could be used in scripts type checking to allow fallback to reading real assemblies on disk like it's done for output assemblies for F# projects currently.

auduchinok commented:
Probably a completely different approach could be used to represent assemblies read by assembly reader shim (like there's a type to represent an F# project) but I found the approach used in this PR to be the least-churning out of options I've been thinking about.

cartermp commented:
@KevinRansom this would be good to look at as a part of your FSI work (once the NuGet plugin and basic IDE hookups are done)

dsyme commented on src/fsharp/CompileOps.fsi:
```diff
@@ -160,8 +161,9 @@ type IRawFSharpAssemblyData =
 
 type TimeStampCache = 
-    new: defaultTimeStamp: DateTime -> TimeStampCache
+    new: defaultTimeStamp: DateTime * [<Optional; DefaultParameterValue(false)>] isInteractive: bool -> TimeStampCache
```
Just make this a non-optional argument, we may as well make it explicit everywhere

dsyme commented on src/absil/ilread.fs:
```diff
@@ -3943,7 +3943,8 @@ type ILReaderOptions =
       metadataOnly: MetadataOnlyFlag
-      tryGetMetadataSnapshot: ILReaderTryGetMetadataSnapshot }
+      tryGetMetadataSnapshot: ILReaderTryGetMetadataSnapshot
+      isInteractive: bool }
```
I think `isInteractive` should be called `requiresOnDiskAssemblies` or `requiresFilesFromRealFileSystem` or `bypassFileSystemShim` or something like that for most of the changes in this PR

auduchinok replied:
Should we introduce a different name in TcConfig/TcConfigBuilder in a separate PR as well? I could try coming up with a more descriptive one, like your examples.

dsyme replied:
I think that could be tricky, there are several implications of the name that would be hard to convey

auduchinok commented:
I've applied review feedback:
* change `isInteractive` to `bypassFileSystemShim`
* make `TimeStampCache` ctor parameter non-optional

In addition to that I've extracted a common shim interface for checking files existence and write timestamp.

I've also updated few other places where the assembly reader shim should be used instead of the file system shim.

auduchinok commented on src/fsharp/SimulatedMSBuildReferenceResolver.fs:
```diff
@@ -102,7 +102,7 @@ let internal SimulatedMSBuildResolver =
                 try
                     if not found && Path.IsPathRooted r then
-                        if FileSystem.SafeExists r then
+                        if AssemblyReader.Exists (r, bypassFileSystemShim) then
```
I'm not sure if this resolver should be updated for the new APIs.
There're few `Directory.Exists` checks that should be updated too if we do it.

KevinRansom commented:
@auduchinok ,

is this still something you wish to pursue, it has been a year since you last re-visited it.  I am trying to remove orphaned PR's from the repo.

Thanks

Kevin

auduchinok commented:
@KevinRansom Sadly, I haven't looked at the corresponding branch in our plugin either. I'll have to revisit it when resurrect that branch, but at this point it'll be fine to reopen the PR or create a new one.
Thank you for cleaning up it all!
